dali/operators/io/file/file_read.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/operator/checkpointing/stateless_operator.h"
+#include "dali/core/util.h"
+#include "dali/util/file.h"
+
+namespace dali {
+namespace io {
+
+class FileRead : public StatelessOperator<CPUBackend> {
+ public:
+  explicit FileRead(const OpSpec &spec)
+      : StatelessOperator<CPUBackend>(spec),
+        dont_use_mmap_(spec.GetArgument<bool>("dont_use_mmap")),
+        use_o_direct_(spec.GetArgument<bool>("use_o_direct")) {}
+
+  bool CanInferOutputs() const override { return true; }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_descs, const Workspace &ws) override {
+    const auto &filepaths = ws.Input<CPUBackend>(0);
+    DALI_ENFORCE(filepaths.type() == DALI_UINT8 || filepaths.type() == DALI_INT8,
+                 "Unexpected input data type");
+    auto nsamples = filepaths.shape().num_samples();
+    output_descs.resize(1);
+    output_descs[0].type = DALI_UINT8;
+    output_descs[0].shape.resize(nsamples, 1);
+    filenames_.resize(nsamples);
+    files_.resize(nsamples);
+    for (int i = 0; i < nsamples; i++) {
+      size_t filename_len = filepaths.tensor_shape_span(i)[0];
+      filenames_[i].resize(filename_len + 1);
+      std::memcpy(&filenames_[i][0], filepaths.raw_tensor(i), filename_len);
+      filenames_[i][filename_len] = '\0';
+
+      FileStream::Options opts;
+      opts.use_mmap = !dont_use_mmap_;
+      opts.use_odirect = use_o_direct_;
+      opts.read_ahead = false;
+      files_[i] = FileStream::Open(filenames_[i], opts);
+      output_descs[0].shape.tensor_shape_span(i)[0] = files_[i]->Size();
+    }
```

Comment:
Is opening a file a possible bottleneck? I think you may have a bit of expertise in that topic based on the imgcodec work?

Would there be a benefit in running that in a thread pool or would it cause more harm?

Reply:
There's no such thing as opening the file in S3. However, querying the file size involves a server request, so we might want to parallelize it. Good point.